_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,323 @@
+                producer.send(new ProducerRecord<>(TOPIC_NAME, messageCount.get(), value), this);
+                messageCount.incrementAndGet();
+            }
+        }
+    }
+
+    private KafkaProducer<Long, byte[]> createKafkaProducer() {
+        // Create properties for the Kafka producer
+        Properties props = new Properties();
+        
+        // Configure the connection to Kafka brokers
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
+
+        // Set a unique client ID for tracking
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "client-" + UUID.randomUUID());
+        
+        // Configure serializers for keys and values
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, LongSerializer.class);
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+
+        return new KafkaProducer<>(props);
+    }
+
+    private void sleep(long ms) {
+        try {
+            TimeUnit.MILLISECONDS.sleep(ms);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private byte[] randomBytes(int size) {
+        if (size <= 0) {
+            throw new IllegalArgumentException("Record size must be greater than zero");
+        }
+        byte[] payload = new byte[size];
+        for (int i = 0; i < payload.length; ++i) {
+            payload[i] = (byte) (RND.nextInt(26) + 65);
+        }
+        return payload;
+    }
+
+    private boolean retriable(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (e instanceof IllegalArgumentException
+                || e instanceof UnsupportedOperationException
+                || !(e instanceof RetriableException)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public void onCompletion(RecordMetadata metadata, Exception e) {
+        if (e != null) {
+            // If an exception occurred while sending the record
+            System.err.println(e.getMessage());
+
+            if (!retriable(e)) {
+                // If the exception is not retriable, print the stack trace and exit
+                e.printStackTrace();
+                System.exit(1);
+            }
+        } else {
+            // If the record was successfully sent
+            System.out.printf("Record sent to %s-%d with offset %d%n",
+                    metadata.topic(), metadata.partition(), metadata.offset());
+        }
+    }
+}
+```
+
+### Running the producer client
+
+To put this client into action, simply run the main method in the `Producer` class. 
+When running, it creates the message payloads using randomly generated byte arrays. 
+The client produces messages until it reaches the predefined message count, which is 100 messages with the `NUM_MESSAGES` constant value we specified. 
+
+> With its thread-safe design, multiple threads can share a single producer instance.
```

Comment:
Maybe link to docs that attest of this statement?

Reply:
Not sure what docs to link to, so changed the note!

> Kafka producer instances are designed to be thread-safe, allowing multiple threads to share a single producer instances.

Reply:
In case you want a source to reference, the [KafkaProducer Javadoc](https://kafka.apache.org/35/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html) states `The producer is thread safe and sharing a single producer instance across threads will generally be faster than having multiple instances. `